src/internal/connector/onedrive/restore.go

```diff
@@ -118,50 +146,141 @@ func RestoreCollection(
 		select {
 		case <-ctx.Done():
 			errUpdater("context canceled", ctx.Err())
-			return metrics, true
+			return metrics, folderPerms, true
 
 		case itemData, ok := <-items:
 			if !ok {
-				return metrics, false
-			}
-			metrics.Objects++
-
-			metrics.TotalBytes += int64(len(copyBuffer))
-
-			itemInfo, err := restoreItem(ctx,
-				service,
-				itemData,
-				drivePath.DriveID,
-				restoreFolderID,
-				copyBuffer,
-				source)
-			if err != nil {
-				errUpdater(itemData.UUID(), err)
-				continue
+				return metrics, folderPerms, false
 			}
 
 			itemPath, err := dc.FullPath().Append(itemData.UUID(), true)
 			if err != nil {
 				logger.Ctx(ctx).DPanicw("transforming item to full path", "error", err)
 				errUpdater(itemData.UUID(), err)
-
 				continue
 			}
 
-			deets.Add(
-				itemPath.String(),
-				itemPath.ShortRef(),
-				"",
-				true,
-				itemInfo)
+			if source == OneDriveSource {
+				name := itemData.UUID()
+				if strings.HasSuffix(name, DataFileSuffix) {
+					metrics.Objects++
+					metrics.TotalBytes += int64(len(copyBuffer))
+					trimmedName := strings.TrimSuffix(name, DataFileSuffix)
+
+					itemID, itemInfo, err = restoreData(ctx, service, trimmedName, itemData, drivePath.DriveID, restoreFolderID, copyBuffer, source)
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					restoredIDs[trimmedName] = itemID
+
+					deets.Add(itemPath.String(), itemPath.ShortRef(), "", true, itemInfo)
+				} else if strings.HasSuffix(name, MetaFileSuffix) {
+					meta, err := getMetadata(itemData.ToReader())
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					trimmedName := strings.TrimSuffix(name, MetaFileSuffix)
+					restoreID, ok := restoredIDs[trimmedName]
+					if !ok {
+						errUpdater(itemData.UUID(), fmt.Errorf("item not available to restore permissions"))
+						continue
+					}
+
+					err = restorePermissions(ctx, service, drivePath.DriveID, restoreID, parentPerms, meta.Permissions)
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					// Objects count is incremented when we restore a
+					// data file and success count is incremented when
+					// we restore a meta file as every data file
+					// should have an associated meta file
+					metrics.Successes++
+				} else if strings.HasSuffix(name, DirMetaFileSuffix) {
+					meta, err := getMetadata(itemData.ToReader())
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					trimmedName := strings.TrimSuffix(name, DirMetaFileSuffix)
+					_, err = createRestoreFolder(ctx, service, drivePath.DriveID, trimmedName, restoreFolderID, parentPerms, meta.Permissions)
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					ipth := itemPath.String()
+					folderPerms[ipth[:len(ipth)-8]] = meta.Permissions
```

Comment:
nit: should this be `timmedName`? (random-looking constants make me nervous 😅)

Reply:
This is the path instead of just the name, let me create a `trimmedPath` variable.